File: C++/ValidAnagram.cpp
#include <iostream>
using namespace std;

class Solution {
public:
    bool isAnagram(string s, string t) {
        int map[26] = {0};
        if (s.length() != t.length()) {
            return false;
        }
        for (int i = 0; i < s.length(); i++) {
            map[s[i] - 'a']++;
        }
        for (int i = 0; i < t.length(); i++) {
            map[t[i] - 'a']--;
        }
        for (int i = 0; i < 26; i++) {  
            if (map[i] != 0) {
                return false;
            }
        }
        return true;
    }
};

int main() {
    return 0;
}

File: Python/valid_anagram.py
# Optimised O(N) time complexity: 2 passes
class Solution(object):
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        freq = {}
        for ptr in range(len(s)):
            freq[s[ptr]] = freq.get(s[ptr], 0) + 1
            freq[t[ptr]] = freq.get(t[ptr], 0) - 1
        for c in freq:
            if freq[c] != 0:
                return False
        return True


# L this is O(N^2) time complexity
class Solution(object):
    def isAnagram(self, s, t):
        if len(s) != len(t):
            return False
        for char in set(s):
            if s.count(char) != t.count(char):
                return False
        return True
